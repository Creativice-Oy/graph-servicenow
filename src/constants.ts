import { RelationshipClass } from '@jupiterone/integration-sdk-core';

export const Steps = {
  USERS: 'step-users',
  GROUPS: 'step-groups',
  GROUP_MEMBERS: 'step-group-members',
};

export const Entities = {
  USER: {
    resourceName: 'User',
    _type: 'service_now_user',
    _class: 'User',
  },
  GROUP: {
    resourceName: 'User Group',
    _type: 'service_now_group',
    _class: 'UserGroup',
  },
};

export const Relationships = {
  GROUP_HAS_USER: {
    _type: 'service_now_group_has_user',
    sourceType: Entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER._type,
  },
  GROUP_HAS_GROUP: {
    _type: 'service_now_group_has_group',
    sourceType: Entities.GROUP._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.GROUP._type,
  },
};
