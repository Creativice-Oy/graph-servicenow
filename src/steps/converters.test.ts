import {
  createUserEntity,
  createGroupEntity,
  createGroupUserRelationship,
  createGroupGroupRelationship,
} from './converters';
import { Entities } from '../constants';

test('createUserEntity', () => {
  const user = {
    calendar_integration: '1',
    country: '',
    user_password: 'some-bogus-password',
    last_login_time: '2020-09-23 22:00:16',
    source: '',
    sys_updated_on: '2020-09-23 22:07:54',
    building: '',
    web_service_access_only: 'false',
    notification: '2',
    enable_multifactor_authn: 'false',
    sys_updated_by: 'j1-administrator',
    sys_created_on: '2020-09-21 17:57:17',
    sys_domain: {
      link:
        'https://dev94579.service-now.com/api/now/table/sys_user_group/global',
      value: 'global',
    },
    state: '',
    vip: 'false',
    sys_created_by: 'admin',
    zip: '',
    home_phone: '',
    time_format: '',
    last_login: '2020-09-23',
    default_perspective: '',
    active: 'true',
    sys_domain_path: '/',
    cost_center: '',
    phone: '',
    name: 'Jupiterone Administrator',
    employee_number: '',
    password_needs_reset: 'false',
    gender: '',
    city: '',
    failed_attempts: '0',
    user_name: 'j1-administrator',
    roles: '',
    title: '',
    sys_class_name: 'sys_user',
    sys_id: '5ccbba87db1310109d87a9954b9619db',
    internal_integration_user: 'false',
    ldap_server: '',
    mobile_phone: '',
    street: '',
    company: '',
    department: {
      link:
        'https://dev94579.service-now.com/api/now/table/cmn_department/9d7802bfdb1710109d87a9954b96199d',
      value: '9d7802bfdb1710109d87a9954b96199d',
    },
    first_name: 'Jupiterone',
    email: 'j1-admin@jupiterone.com',
    introduction: '',
    preferred_language: '',
    manager: '',
    locked_out: 'false',
    sys_mod_count: '5',
    last_name: 'Administrator',
    photo: '',
    avatar: '',
    middle_name: '',
    sys_tags: '',
    time_zone: '',
    schedule: '',
    date_format: '',
    location: '',
  };

  const userEntity = createUserEntity(user);

  expect(userEntity).toMatchSnapshot();
  expect(userEntity).toMatchGraphObjectSchema({
    _class: Entities.USER._class,
    schema: {},
  });
});

test('createGroupEntity', () => {
  const group = {
    parent: {
      link:
        'https://dev94579.service-now.com/api/now/table/sys_user_group/43f6027fdb1710109d87a9954b961939',
      value: '43f6027fdb1710109d87a9954b961939',
    },
    manager: '',
    roles: '',
    sys_mod_count: '0',
    active: 'true',
    description: '',
    source: '',
    sys_updated_on: '2020-09-23 22:03:48',
    sys_tags: '',
    type: '',
    sys_id: '66a74a7fdb1710109d87a9954b961914',
    sys_updated_by: 'j1-administrator',
    cost_center: '',
    default_assignee: '',
    sys_created_on: '2020-09-23 22:03:48',
    name: 'j1-subgroup',
    exclude_manager: 'false',
    email: 'j1-subgroup@jupiterone.com',
    include_members: 'false',
    sys_created_by: 'j1-administrator',
  };

  const groupEntity = createGroupEntity(group);

  expect(groupEntity).toMatchSnapshot();
  expect(groupEntity).toMatchGraphObjectSchema({
    _class: Entities.GROUP._class,
    schema: {},
  });

  const groupGroupRelationship = createGroupGroupRelationship(
    groupEntity,
    group.parent,
  );
  expect(groupGroupRelationship).toMatchSnapshot();
});

test('createGroupUserRelationship', () => {
  const groupUser = {
    sys_id: '94878e3fdb1710109d87a9954b96195f',
    sys_updated_by: 'j1-administrator',
    sys_created_on: '2020-09-23 22:02:55',
    sys_mod_count: '0',
    sys_updated_on: '2020-09-23 22:02:55',
    sys_tags: '',
    user: {
      link:
        'https://dev94579.service-now.com/api/now/table/sys_user/5ccbba87db1310109d87a9954b9619db',
      value: '5ccbba87db1310109d87a9954b9619db',
    },
    sys_created_by: 'j1-administrator',
    group: {
      link:
        'https://dev94579.service-now.com/api/now/table/sys_user_group/43f6027fdb1710109d87a9954b961939',
      value: '43f6027fdb1710109d87a9954b961939',
    },
  };

  const userGroupRelationship = createGroupUserRelationship(groupUser);

  expect(userGroupRelationship).toMatchSnapshot();
});
