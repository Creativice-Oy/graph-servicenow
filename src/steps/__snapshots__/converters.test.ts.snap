// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createGroupEntity 1`] = `
Object {
  "_class": Array [
    "UserGroup",
  ],
  "_key": "66a74a7fdb1710109d87a9954b961914",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "active": "true",
        "cost_center": "",
        "createdOn": 1600913028000,
        "default_assignee": "",
        "description": "",
        "email": "j1-subgroup@jupiterone.com",
        "exclude_manager": "false",
        "id": "66a74a7fdb1710109d87a9954b961914",
        "include_members": "false",
        "manager": "",
        "name": "j1-subgroup",
        "parent": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user_group/43f6027fdb1710109d87a9954b961939",
          "value": "43f6027fdb1710109d87a9954b961939",
        },
        "roles": "",
        "source": "",
        "sys_created_by": "j1-administrator",
        "sys_created_on": "2020-09-23 22:03:48",
        "sys_id": "66a74a7fdb1710109d87a9954b961914",
        "sys_mod_count": "0",
        "sys_tags": "",
        "sys_updated_by": "j1-administrator",
        "sys_updated_on": "2020-09-23 22:03:48",
        "type": "",
        "updatedOn": 1600913028000,
      },
    },
  ],
  "_type": "service_now_group",
  "active": true,
  "createdOn": 1600913028000,
  "description": "",
  "displayName": "j1-subgroup",
  "email": "j1-subgroup@jupiterone.com",
  "id": "66a74a7fdb1710109d87a9954b961914",
  "name": "j1-subgroup",
  "updatedOn": 1600913028000,
}
`;

exports[`createGroupEntity 2`] = `
Object {
  "_class": "HAS",
  "_fromEntityKey": "43f6027fdb1710109d87a9954b961939",
  "_key": "43f6027fdb1710109d87a9954b961939|has|66a74a7fdb1710109d87a9954b961914",
  "_toEntityKey": "66a74a7fdb1710109d87a9954b961914",
  "_type": "service_now_group_has_group",
  "displayName": "HAS",
}
`;

exports[`createGroupUserRelationship 1`] = `
Object {
  "_class": "HAS",
  "_fromEntityKey": "43f6027fdb1710109d87a9954b961939",
  "_key": "43f6027fdb1710109d87a9954b961939|has|5ccbba87db1310109d87a9954b9619db",
  "_toEntityKey": "5ccbba87db1310109d87a9954b9619db",
  "_type": "service_now_group_has_user",
  "createdOn": 1600912975000,
  "displayName": "HAS",
  "id": "94878e3fdb1710109d87a9954b96195f",
  "updatedOn": 1600912975000,
}
`;

exports[`createUserEntity 1`] = `
Object {
  "_class": Array [
    "User",
  ],
  "_key": "5ccbba87db1310109d87a9954b9619db",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "active": "true",
        "avatar": "",
        "building": "",
        "calendar_integration": "1",
        "city": "",
        "company": "",
        "cost_center": "",
        "country": "",
        "createdOn": 1600725437000,
        "date_format": "",
        "default_perspective": "",
        "department": Object {
          "link": "https://dev94579.service-now.com/api/now/table/cmn_department/9d7802bfdb1710109d87a9954b96199d",
          "value": "9d7802bfdb1710109d87a9954b96199d",
        },
        "email": "j1-admin@jupiterone.com",
        "employee_number": "",
        "enable_multifactor_authn": "false",
        "failed_attempts": "0",
        "first_name": "Jupiterone",
        "gender": "",
        "home_phone": "",
        "id": "5ccbba87db1310109d87a9954b9619db",
        "internal_integration_user": "false",
        "introduction": "",
        "last_login": "2020-09-23",
        "last_login_time": "2020-09-23 22:00:16",
        "last_name": "Administrator",
        "ldap_server": "",
        "location": "",
        "locked_out": "false",
        "manager": "",
        "middle_name": "",
        "mobile_phone": "",
        "name": "Jupiterone Administrator",
        "notification": "2",
        "password_needs_reset": "false",
        "phone": "",
        "photo": "",
        "preferred_language": "",
        "roles": "",
        "schedule": "",
        "source": "",
        "state": "",
        "street": "",
        "sys_class_name": "sys_user",
        "sys_created_by": "admin",
        "sys_created_on": "2020-09-21 17:57:17",
        "sys_domain": Object {
          "link": "https://dev94579.service-now.com/api/now/table/sys_user_group/global",
          "value": "global",
        },
        "sys_domain_path": "/",
        "sys_id": "5ccbba87db1310109d87a9954b9619db",
        "sys_mod_count": "5",
        "sys_tags": "",
        "sys_updated_by": "j1-administrator",
        "sys_updated_on": "2020-09-23 22:07:54",
        "time_format": "",
        "time_zone": "",
        "title": "",
        "updatedOn": 1600913274000,
        "user_name": "j1-administrator",
        "user_password": "[REDACTED]",
        "vip": "false",
        "web_service_access_only": "false",
        "zip": "",
      },
    },
  ],
  "_type": "service_now_user",
  "active": true,
  "createdOn": 1600725437000,
  "displayName": "Jupiterone Administrator",
  "email": "j1-admin@jupiterone.com",
  "id": "5ccbba87db1310109d87a9954b9619db",
  "name": "Jupiterone Administrator",
  "updatedOn": 1600913274000,
  "username": "j1-administrator",
}
`;
